### PR TITLE
Prefill kvcache upstream

### PR DESCRIPTION
**Pull request by puneeshkhanna**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by puneeshkhanna** [comment]
@regisss, @libinta, @dvarshney-habana  - Please add 1.16 synpase release label to this.

**Comment by puneeshkhanna** [comment]
Description of the changes in this PR -

Pad KV cache to full input + new tokens len for decode phase. Delete the KV cache used as inputs by HPU graphs after full prompt generation. Ensure KV cache is not returned as output tensor during decode phase. Deletion of KV cache input tensor used by HPU graphs needs to be protected by PT_HPUGRAPH_DISABLE_TENSOR_CACHE env variable. All the changes are protected by bucket internal flag right now.


Updated command (remove --reuse_cache from all existing commands , setting PT_HPUGRAPH_DISABLE_TENSOR_CACHE=1 automatically taken care)
python ../gaudi_spawn.py --use_deepspeed --world_size 8 run_generation.py --model_name_or_path /mnt/weka/data/pytorch/llama2/Llama-2-70b-hf/ --use_hpu_graphs --use_kv_cache --max_input_tokens 2048 --max_new_tokens 2048 --batch_size 200 --attn_softmax_bf16 --trim_logits --bf16 --warmup 2 --n_iterations 2 --limit_hpu_graphs --bucket_internal --bucket_size 128

With the changes in this PR, performance in any existing configs remains same but we can scale batch sizes to much much higher numbers since we save a lot of memory during the prefill phase.
As an example with 2K input tokens + 2K new tokens, llama 70B on 8x with flash attention - maximum batch size without PR changes that we can go is around 270 and with the changes in this PR, we can go up to batch size 370.

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -380,11 +380,11 @@ def pre_attn_forward(
                 past_key_value = (self.k_cache.get_shape(), self.v_cache.get_shape())
             else:
                 if past_key_value is None:
-                    past_key = torch.zeros(key_states.shape, dtype=self.k_proj.weight.dtype, device=key_states.device)
                     past_value = torch.zeros(
                         key_states.shape, dtype=self.k_proj.weight.dtype, device=key_states.device
                     )
-                    past_key_value = (past_key, past_value)
+                    # Return list instead of tuple
+                    past_key_value = [past_key, past_value]
```
Comment (libinta) [inline optimum/habana/transformers/models/llama/modeling_llama.py:387]: this could impact tgi as tgi goes through this route

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -380,11 +380,11 @@ def pre_attn_forward(
                 past_key_value = (self.k_cache.get_shape(), self.v_cache.get_shape())
             else:
                 if past_key_value is None:
-                    past_key = torch.zeros(key_states.shape, dtype=self.k_proj.weight.dtype, device=key_states.device)
                     past_value = torch.zeros(
                         key_states.shape, dtype=self.k_proj.weight.dtype, device=key_states.device
                     )
```
Comment (ssarkar2) [inline optimum/habana/transformers/models/llama/modeling_llama.py:385]: only past_value is defined, past_key is not.

it needs to be dfined like here: https://github.com/HabanaAI/optimum-habana-fork/blob/habana-main/optimum/habana/transformers/models/llama/modeling_llama.py#L304

**Comment by ssarkar2** [comment]
merged thru: https://github.com/huggingface/optimum-habana/pull/1028
